src/v/kafka/server/handlers/produce.cc

```diff
@@ -279,37 +279,104 @@ static partition_produce_stages produce_topic_partition(
                         .partition_index = ntp.tp.partition,
                         .error_code = error_code::not_leader_for_partition});
                 }
-                auto stages = partition_append(
-                  ntp.tp.partition,
-                  ss::make_lw_shared<replicated_partition>(
-                    std::move(partition)),
-                  bid,
-                  std::move(reader),
-                  acks,
-                  num_records);
-                return stages.dispatched
-                  .then_wrapped([source_shard, dispatch = std::move(dispatch)](
-                                  ss::future<> f) mutable {
-                      if (f.failed()) {
-                          (void)ss::smp::submit_to(
-                            source_shard,
-                            [dispatch = std::move(dispatch),
-                             e = f.get_exception()]() mutable {
-                                dispatch->set_exception(e);
-                                dispatch.reset();
+                auto f = ss::now();
+                if (partition->is_compacted() && batch.compressed()) {
+                    batch.key_offsets(std::vector<model::key_offset>());
+                    f = storage::internal::decompress_batch(batch).then(
+                      [&batch](model::record_batch&& b) {
+                          return model::for_each_record(
+                            b, [&batch](const model::record& r) {
+                                batch.key_offsets()->push_back(
+                                  model::key_offset(
+                                    r.offset_delta(),
+                                    r.key_size(),
+                                    iobuf_to_bytes(r.key())));
                             });
-                          return;
-                      }
-                      (void)ss::smp::submit_to(
-                        source_shard,
-                        [dispatch = std::move(dispatch)]() mutable {
-                            dispatch->set_value();
-                            dispatch.reset();
-                        });
-                  })
-                  .then([f = std::move(stages.produced)]() mutable {
-                      return std::move(f);
-                  });
+                      });
+                }
+                return f.then([partition = std::move(partition),
```

Comment:
should we make some of these `.then()` separate free functions. Seems hard to read at a glance due to future nesting.